proposals/type-classes.md

```diff
@@ -237,60 +202,155 @@ fun addInts(a: Int, b: Int, with Monoid<Int>): Int = add(a, b)
+
+To determine whether a typeclass definition is a valid type-side implementation we perform the following check:
+
+1. A "local type" is any type (but not typealias) defined in the current file (e.g. everything defined in `data.bar` if we're evaluating `data.bar`).
+2. A generic type parameter is "covered" by a type if it occurs within that type, e.g. `MyType` covers `T` in `MyType<T>` but not `Pair<T, MyType>`.
+3. Write out the parameters to the type class in order.
+4. The parameters must include a type defined in this file.
+5. Any generic type parameters must occur after the first instance of a local type or be covered by a local type.
+
+If a type class implementation meets these rules it is a valid type-side implementation.
+ 
+
+## Compile Resolution Rules
+
+When the compiler finds a call site invoking a function that has type class instances constraints declared with `with` as in the example below:
+
+Declaration: 
+```kotlin
+fun <A> add(a: A, b: A, with Monoid<A>): A = a.combine(b)
+```
+Call site:
 ```kotlin
-extension object AddingInts {
-  fun addInts(): Int = add(1, 2)
+fun addInts(): Int = add(1, 2)
+```
+
+1. The compiler first looks at the file the interface is defined in. If it finds exactly one implementation it uses that instance.
+2. If it fails to find an implementation in the interface's file, it then looks at the files of the implemented types. For each type class parameter check the file it was defined in. If exactly one implementation is found use that instance.
+3. If no matching implementation is found in either of these places fail to compile.
+4. If more than one matching implementation is found, fail to compile and indicate that there or conflicting instances.
+
+Some of these examples where originally proposed by Roman Elizarov and the Kategory contributors where these features where originally discussed https://github.com/Kotlin/KEEP/pull/87
+
+## Appendix A: Orphan Implementations
+
+Orphan implementations are a subject of controversy. Combining two libraries, one defining a data type, the other defining an interface, is a feature that many programmers have longed for. However, implementing this feature in a way that doesn't break other features of interfaces is difficult and drastically complicates how the compiler works with those interfaces.
+
+Orphan implementations are the reason that type classes have often been described as "anti-modular" as the most common way of dealing with them is through global coherency checks. This is necessary to ensure that two libraries have not defined incompatible implementations of a type class interface.
+
+Relaxing the orphan rules is a backwards-compatible change. If this proposal is accepted without permitting orphans it is useful to consider how they could be added in the future.
```

Comment:
IMO The issue of not considering internal overrides here is that it makes this proposal less useful to the OOP Kotlin community

Reply:
The type-class instance of a subclass can still differ from a superclass, this doesn't preclude that. I think piling on overrides throughout the feature without knowing how people will use it in practice is dangerous and V1 of type classes should be the minimal set of features that we know people want.